### PR TITLE
Added config, raid check and random difficulty chance loot conditions

### PR DESCRIPTION
**Pull request by abigailfails**
Finally got around to doing this! The config loot condition was just added because i realised the 'normal' condition type isn't compatible with loot tables, and the random difficulty and raid conditions should be generally useful in the future. I'm also going to need them to convert some hardcoded loot table stuff into fully data-driven modifiers in S&R once those are done :P

I'm still a bit iffy about the documentation, so tell me if you see anything wrong with that.

### CODE REVIEW

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/ConfigLootCondition.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                json.add("predicates", predicates);
+                for (Map.Entry<IConfigPredicate, Boolean> predicatePair : value.predicates.entrySet()) {
+                    IConfigPredicate predicate = predicatePair.getKey();
+                    ResourceLocation predicateID = predicate.getID();
+                    JsonObject object = new JsonObject();
+                    predicates.add(object);
+                    object.addProperty("type", predicateID.toString());
+                    ConfigCondition.Serializer.CONFIG_PREDICATE_SERIALIZERS.get(predicateID).write(object, predicate);
+                    object.addProperty("inverted", predicatePair.getValue());
+                }
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/ConfigLootCondition.java:83]: Small style suggestions. The formatting in this suggestion may be a bit messed up.
```suggestion
                JsonArray predicates = new JsonArray();
                for (Map.Entry<IConfigPredicate, Boolean> predicatePair : value.predicates.entrySet()) {
                    IConfigPredicate predicate = predicatePair.getKey();
                    ResourceLocation predicateID = predicate.getID();
                    JsonObject object = new JsonObject();
                    object.addProperty("type", predicateID.toString());
                    ConfigCondition.Serializer.CONFIG_PREDICATE_SERIALIZERS.get(predicateID).write(object, predicate);
                    object.addProperty("inverted", predicatePair.getValue());
                    predicates.add(object);
                }
                json.add("predicates", predicates);
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/ConfigLootCondition.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            }
+            if (value.inverted) json.addProperty("inverted", true);
+        }
+
+        @Override
+        public ConfigLootCondition deserialize(JsonObject json, JsonDeserializationContext context) {
+            if (!json.has("value"))
+                throw new JsonSyntaxException("Missing 'value', expected to find a string");
+            String name = JSONUtils.getString(json, "value");
+            ForgeConfigSpec.ConfigValue<?> configValue = configValues.get(name);
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/ConfigLootCondition.java:93]:
```suggestion
            ForgeConfigSpec.ConfigValue<?> configValue = this.configValues.get(name);
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RaidCheckCondition.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public boolean test(LootContext lootContext) {
+        Entity entity = lootContext.get(LootParameters.THIS_ENTITY);
+        return inverted != (entity != null && lootContext.getWorld().findRaid(entity.getPosition()) != null);
+    }
+
+    public static class Serializer implements ILootSerializer<RaidCheckCondition> {
+
+        public void serialize(JsonObject json, RaidCheckCondition condition, JsonSerializationContext context) {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RaidCheckCondition.java:45]:
```suggestion
        @Override
        public void serialize(JsonObject json, RaidCheckCondition condition, JsonSerializationContext context) {
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RaidCheckCondition.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            if (condition.inverted)
+                json.addProperty("inverted", true);
+        }
+
+        public RaidCheckCondition deserialize(JsonObject json, JsonDeserializationContext context) {
+            return new RaidCheckCondition(JSONUtils.getBoolean(json, "inverted", false));
+        }
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RaidCheckCondition.java:52]:
```suggestion
        @Override
        public RaidCheckCondition deserialize(JsonObject json, JsonDeserializationContext context) {
            return new RaidCheckCondition(JSONUtils.getBoolean(json, "inverted", false));
        }
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+public class RandomDifficultyChanceCondition implements ILootCondition {
+	private final float defaultChance;
+	private final float peacefulChance;
+	private final float easyChance;
+	private final float normalChance;
+	private final float hardChance;
+
+	public RandomDifficultyChanceCondition(float defaultChance, float peacefulChance, float easyChance, float normalChance, float hardChance) {
+		this.peacefulChance = peacefulChance;
+		this.defaultChance = defaultChance;
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java:40]:
```suggestion
		this.defaultChance = defaultChance;
		this.peacefulChance = peacefulChance;
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+				break;
+			case HARD:
+				if (this.hardChance >= 0) chance = this.hardChance;
+
+		}
+		return lootContext.getRandom().nextFloat() < chance;
+	}
+
+	public static class Serializer implements ILootSerializer<RandomDifficultyChanceCondition> {
+		public void serialize(JsonObject json, RandomDifficultyChanceCondition condition, JsonSerializationContext context) {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java:72]:
```suggestion

		public void serialize(JsonObject json, RandomDifficultyChanceCondition condition, JsonSerializationContext context) {
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+				return new RandomDifficultyChanceCondition(JSONUtils.getFloat(json, "default_chance"), getFloatOrMinus1(json, "peaceful"), getFloatOrMinus1(json, "easy"), getFloatOrMinus1(json, "normal"), getFloatOrMinus1(json, "hard"));
+			}
+			throw new JsonSyntaxException("Missing 'default_chance', expected to find a float");
+		}
+
+		private static float getFloatOrMinus1(JsonObject json, String fieldName) {
+			if (json.has(fieldName))
+				return JSONUtils.getFloat(json, fieldName);
+			return -1;
+		}
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java:95]:
```suggestion
		}

```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			if (json.has("default_chance")) {
+				return new RandomDifficultyChanceCondition(JSONUtils.getFloat(json, "default_chance"), getFloatOrMinus1(json, "peaceful"), getFloatOrMinus1(json, "easy"), getFloatOrMinus1(json, "normal"), getFloatOrMinus1(json, "hard"));
+			}
+			throw new JsonSyntaxException("Missing 'default_chance', expected to find a float");
+		}
+
+		private static float getFloatOrMinus1(JsonObject json, String fieldName) {
+			if (json.has(fieldName))
+				return JSONUtils.getFloat(json, fieldName);
+			return -1;
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/loot/RandomDifficultyChanceCondition.java:94]:
```suggestion
			return json.has(fieldName) ? JSONUtils.getFloat(json, fieldName) : -1.0F;
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import com.minecraftabnormals.abnormals_core.core.AbnormalsCore;
+import com.minecraftabnormals.abnormals_core.core.api.conditions.loot.RaidCheckCondition;
+import com.minecraftabnormals.abnormals_core.core.api.conditions.loot.RandomDifficultyChanceCondition;
+import net.minecraft.loot.LootConditionType;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.registry.Registry;
+
+/**
+ * @author abigailfails
+ * */
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java:12]:
```suggestion
 */
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+public class ACLootConditions {
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java:13]: There could be more documentation about this class's purpose.

Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java:13]:
```suggestion
public final class ACLootConditions {
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 11]
```diff
@@ -299,21 +302,22 @@ public static void registerConfigCondition(String modId, Object... configObjects
 				}
 			}
 		}
-		CraftingHelper.register(new ConfigValueCondition.Serializer(modId, configValues));
+		CraftingHelper.register(new ConfigCondition.Serializer(modId, configValues));
+		Registry.register(Registry.LOOT_CONDITION_TYPE, new ResourceLocation(AbnormalsCore.MODID, "config"), new LootConditionType(new ConfigLootCondition.Serializer(modId, configValues)));
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:306]: Shouldn't this use `modId`, not `AbnormalsCore.MODID`?

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 12]
```diff
@@ -218,14 +219,14 @@ public static void addToJigsawPattern(ResourceLocation toAdd, JigsawPiece newPie
...
 
 	/**
-	 * Registers a {@link ConfigValueCondition.Serializer} under the name {@code "[modId]:config"}
+	 * Registers a {@link ConfigCondition.Serializer} and a {@link ConfigLootCondition.Serializer} under the name {@code "[modId]:config"}
 	 * that accepts the values of {@link ConfigKey} annotations for {@link net.minecraftforge.common.ForgeConfigSpec.ConfigValue}
 	 * fields in the passed-in collection of objects, checking against the annotation's corresponding
 	 * {@link net.minecraftforge.common.ForgeConfigSpec.ConfigValue} to determine whether the condition should pass.<br><br>
 	 * <h2>Function</h2>
-	 * <p>This method allows you to make crafting recipes, advancement modifiers, etc. check whether a specific config
+	 * <p>This method allows you to make crafting recipes, advancement modifiers, loot tables, etc. check whether a specific config
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:227]:
```suggestion
	 * <p>This method allows you to make crafting recipes, modifiers, loot tables, etc. check whether a specific config
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 13]
```diff
@@ -218,14 +219,14 @@ public static void addToJigsawPattern(ResourceLocation toAdd, JigsawPiece newPie
...
 	 * field is true/whether it meets specific predicates before loading without having to hardcode new condition classes
-	 * for certain config values. It's essentially a wrapper for {@link CraftingHelper#register(IConditionSerializer)}
+	 * for certain config values. It's essentially a wrapper for the condition and loot condition registry methods and
 	 * and should be called during common setup accordingly.</p><br><br>
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:230]:
```suggestion
	 * for certain config values. It's essentially a wrapper for the condition and loot condition registry methods and
	 * should be called during common setup accordingly.</p><br><br>
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 14]
```diff
@@ -250,10 +251,12 @@ public static void addToJigsawPattern(ResourceLocation toAdd, JigsawPiece newPie
 	 * ]
 	 * }</pre>
 	 *
-	 * <p>Config conditions also accept a {@code predicates} array which defines predicates the config value must match
-	 * before the condition returns true, and a boolean {@code inverted} argument which makes the condition pass if it
-	 * evaluates to false instead of true. If the config value is non-boolean, {@code predicates} is required.
-	 * Each individual predicate also accepts an {@code inverted} argument, as {@code !(A.B) != !A.!B}.</p>
+	 * <p>Config conditions also accept a {@code predicates} array which defines
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:254]:
```suggestion
	 * <p>Config conditions also accept a {@code predicates} array, which defines
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 15]
```diff
@@ -250,10 +251,12 @@ public static void addToJigsawPattern(ResourceLocation toAdd, JigsawPiece newPie
...
+	 * {@link com.minecraftabnormals.abnormals_core.core.api.conditions.config.IConfigPredicate IConfigPredicate}s the
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:255]:
```suggestion
	 * {@link com.minecraftabnormals.abnormals_core.core.api.conditions.config.IConfigPredicate IConfigPredicate}s that the
```

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 16]
```diff
@@ -250,10 +251,12 @@ public static void addToJigsawPattern(ResourceLocation toAdd, JigsawPiece newPie
...
+	 * config value must match before the condition returns true, and a boolean {@code inverted} argument which makes
+	 * the condition pass if it evaluates to false instead of true. If the config value is non-boolean,
+	 * {@code predicates} is required. Each individual predicate also accepts an {@code inverted} argument (as
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:258]:
```suggestion
	 * {@code predicates} are required. Each individual predicate also accepts an {@code inverted} argument (as
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java (hunk 9 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java:12]: Ugh, those always happen

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java (hunk 11 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:306]: ...yes.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java (hunk 10 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java]: Is what i just committed the right direction?

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java (hunk 11 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java]: Fixed.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/ConfigCondition.java [hunk 17]
```diff
@@ -28,25 +29,24 @@
...
- *   <li>{@code inverted}   - whether the condition should be inverted, so it will pass if {@code predicates} return false instead.
- *                            Inclusion is optional.</li>
+ *   <li>{@code inverted} (optional)   - whether the condition should be inverted, so it will pass if {@code predicates} return false instead.</li>
  * </ul></p>
  *
  * @see DataUtil#registerConfigCondition(String, Object...)
  * @author abigailfails
  */
-public class ConfigValueCondition implements ICondition {
+public class ConfigCondition implements ICondition {
```
[the remark this answers is not in the document]
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/ConfigCondition.java]: Reversed it.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/ConfigCondition.java [hunk 18]
```diff
@@ -19,7 +19,8 @@
 import java.util.Map;
 
 /**
- * A condition that checks against the values of config values annotated with {@link ConfigKey}.
+ * A condition that checks against the values of config values annotated with {@link ConfigKey}. Uses the recipe system,
+ * but is also compatible with advancement modifiers etc.
```
[the remark this answers is not in the document]
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/api/conditions/ConfigCondition.java]: Added in latest commit.

File: src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java [hunk 19]
```diff
@@ -1,7 +1,9 @@
 package com.minecraftabnormals.abnormals_core.core.util;
 
+import com.minecraftabnormals.abnormals_core.core.AbnormalsCore;
```
Comment (SmellyModder) [inline src/main/java/com/minecraftabnormals/abnormals_core/core/util/DataUtil.java:3]: Unused import.
```suggestion
```

**Review by SmellyModder — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java (hunk 10 above)
Comment (abigailfails) [reply src/main/java/com/minecraftabnormals/abnormals_core/core/registry/ACLootConditions.java]: Changed it again based on your suggestion.

**Comment by abigailfails** [comment]
when the import is unused! 😳

**Review by SmellyModder — APPROVED** [review]

**Pull request merged**